reference/src/taimpl/test/ta_sa_svp_crypto.cpp

```diff
@@ -0,0 +1,496 @@
+            std::shared_ptr<void>& parameters,
+            std::vector<uint8_t>& iv,
+            std::vector<uint8_t>& counter) {
+
+        switch (cipher_algorithm) {
+            case SA_CIPHER_ALGORITHM_AES_CBC:
+            case SA_CIPHER_ALGORITHM_AES_CBC_PKCS7: {
+                iv = random(AES_BLOCK_SIZE);
+                auto* cipher_parameters_aes_cbc = new sa_cipher_parameters_aes_cbc;
+                cipher_parameters_aes_cbc->iv = iv.data();
+                cipher_parameters_aes_cbc->iv_length = iv.size();
+                parameters = std::shared_ptr<void>(cipher_parameters_aes_cbc);
+                break;
+            }
+            case SA_CIPHER_ALGORITHM_AES_CTR: {
+                iv = random(AES_BLOCK_SIZE);
+                auto* cipher_parameters_aes_ctr = new sa_cipher_parameters_aes_ctr;
+                cipher_parameters_aes_ctr->ctr = iv.data();
+                cipher_parameters_aes_ctr->ctr_length = iv.size();
+                parameters = std::shared_ptr<void>(cipher_parameters_aes_ctr);
+                break;
+            }
+            case SA_CIPHER_ALGORITHM_CHACHA20: {
+                iv = random(CHACHA20_NONCE_LENGTH);
+                counter = {1, 0, 0, 0};
+                auto* cipher_parameters_chacha20 = new sa_cipher_parameters_chacha20;
+                cipher_parameters_chacha20->nonce = iv.data();
+                cipher_parameters_chacha20->nonce_length = iv.size();
+                cipher_parameters_chacha20->counter = counter.data();
+                cipher_parameters_chacha20->counter_length = counter.size();
+                parameters = std::shared_ptr<void>(cipher_parameters_chacha20);
+                break;
+            }
+            default:
+                parameters = nullptr;
+        }
+    }
+
+    size_t get_required_length(
+            sa_cipher_algorithm cipher_algorithm,
+            sa_cipher_mode cipher_mode,
+            size_t key_length,
+            size_t bytes_to_process) {
+
+        switch (cipher_algorithm) {
+            case SA_CIPHER_ALGORITHM_AES_CBC:
+            case SA_CIPHER_ALGORITHM_AES_CTR:
+            case SA_CIPHER_ALGORITHM_AES_ECB:
+            case SA_CIPHER_ALGORITHM_AES_GCM:
+            case SA_CIPHER_ALGORITHM_CHACHA20:
+            case SA_CIPHER_ALGORITHM_CHACHA20_POLY1305:
+                return bytes_to_process;
+
+            case SA_CIPHER_ALGORITHM_AES_ECB_PKCS7:
+            case SA_CIPHER_ALGORITHM_AES_CBC_PKCS7:
+                if (cipher_mode == SA_CIPHER_MODE_ENCRYPT)
+                    return PADDED_SIZE(bytes_to_process);
+                else
+                    return bytes_to_process;
+
+            case SA_CIPHER_ALGORITHM_RSA_PKCS1V15:
+            case SA_CIPHER_ALGORITHM_RSA_OAEP:
+            case SA_CIPHER_ALGORITHM_EC_ELGAMAL:
+                return key_length;
+
+            default:
+                return 0;
+        }
+    }
+
+    bool verify(
+            sa_buffer* buffer,
```

Comment:
Why does the first argument start on the next line?

Reply:
That is the coding standard used across all of secapi3. With multiple parameters, each parameter is on its own line. It just looks funny here because the function name is short.